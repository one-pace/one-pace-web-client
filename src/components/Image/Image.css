.root {
  position: relative;
}

.errorIcon {
  color: #e0e0e0;
  font-size: 48px !important;
  height: 48px;
  width: 48px;
}

.image {
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  left: 0;
}

.container {
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  left: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  pointer-events: none;
}

.spinner {
  display: inline-block;
  height: 48px;
  position: relative;
  width: 48px;

  > svg {
    margin: 0;
  }
}

:global(.md-circular-progress-path) {
  stroke: #ffad6c;
  stroke: var(--lodus-orange);
}
