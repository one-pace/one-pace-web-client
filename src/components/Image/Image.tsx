import React from 'react';
import cn from 'classnames';
// import { CircularProgress, FontIcon } from 'react-md';

import s from './Image.css';

interface Image {
  defaultProps: Props;
  image: any;
  loaderId: string;
  toggle: any;
}

interface Props extends React.ImgHTMLAttributes<HTMLImageElement> {
  /** Alt attribute to show placeholder text */
  // alt: string;
  /** Duration of the fading animation, in milliseconds. */
  animationDuration?: number;
  /** Override aspect ratio. */
  aspectRatio?: number;
  /** Override the default <img> and put any children tags in place of it */
  children?: React.ReactNode;
  /** Overrides the default className */
  className?: string;
  /** Override the background color. */
  color?: string;
  /** Override the crossorigin attribute */
  crossOrigin?: 'anonymous' | 'use-credentials';
  /** Disables the error icon if set to true. */
  disableError?: boolean;
  /** Disables the loading spinner if set to true. */
  disableSpinner?: boolean;
  /** Disables the transition after load if set to true. */
  disableTransition?: boolean;
  /** Override the error icon. */
  errorIcon?: React.ReactNode;
  /** Override the inline-styles of the image. */
  imageStyle?: object;
  /** Override the loading component. */
  // loading?: React.ReactNode;
  /** Fired when the user clicks on the image happened. */
  onClick?: (event?: React.MouseEvent<HTMLDivElement>) => void;
  /** Fired when the image failed to load. */
  onError?: (event?: React.SyntheticEvent<HTMLImageElement>) => void;
  /** Fired when the image finished loading. */
  onLoad?: (event?: React.SyntheticEvent<HTMLImageElement>) => void;
  /** Specifies the URL of an image. */
  src: string;
  /** Override the inline-styles of the root element. */
  style?: object;
}

interface ChildProps {
  /** Alt attribute to show placeholder text */
  alt?: string;
  /** Duration of the fading animation, in milliseconds. */
  animationDuration?: number;
  /** Override aspect ratio. */
  aspectRatio?: number;
  /** Override the default <img> and put any children tags in place of it */
  children?: React.ReactNode;
  /** Overrides the default className */
  className?: string;
  /** Override the background color. */
  color?: string;
  /** Override the crossorigin attribute */
  crossOrigin?: 'anonymous' | 'use-credentials';

  decoding?: 'async';
  /** Disables the error icon if set to true. */
  disableError?: boolean;
  /** Disables the loading spinner if set to true. */
  disableSpinner?: boolean;
  /** Disables the transition after load if set to true. */
  disableTransition?: boolean;
  /** Override the error icon. */
  errorIcon?: React.ReactNode;
  /** Override the inline-styles of the image. */
  imageStyle?: object;
  /** Override the loading component. */
  loading?: React.ReactNode;
  /** Fired when the user clicks on the image happened. */
  onClick?: (event?: React.MouseEvent<HTMLDivElement>) => void;
  /** Fired when the image failed to load. */
  onError?: (event?: React.SyntheticEvent<HTMLImageElement>) => void;
  /** Fired when the image finished loading. */
  onLoad?: (event?: React.SyntheticEvent<HTMLImageElement>) => void;

  ref?: React.Ref<Image>;
  /** Specifies the URL of an image. */
  src?: string;
  /** Override the inline-styles of the root element. */
  style?: object;

  toggle?: any;
}

/**
 * Images are ugly until they're loaded. Materialize it with material image! It will fade in like the material image loading pattern suggests.
 * @see [Image loading patterns](https://material.io/guidelines/patterns/loading-images.html)
 * Based on https://github.com/TeamWertarbyte/material-ui-image/
 */
class Image extends React.PureComponent<Props, any> {
  static defaultProps: {
    animationDuration: number;
    aspectRatio: number;
    children: any;
    color: string;
    crossOrigin: string;
    disableError: boolean;
    disableSpinner: boolean;
    disableTransition: boolean;
    // errorIcon: <FontIcon className={s.errorIcon}>broken_image</FontIcon>,
    errorIcon: any;
    imageStyle: {};
    loading: any;
    onClick: any;
    style: {};
  };

  constructor(props) {
    super(props);
    this.image = React.createRef();
    const srcFileName = props.src.lastIndexOf('//') + 1;
    this.loaderId = props.src.substr(srcFileName);
    this.state = {
      imageError: false,
      imageLoaded: false,
      src: props.src,
    };
  }

  componentDidMount() {
    const img = this.image.current;
    if (img && img.complete) this.handleLoadImage();
  }

  static getDerivedStateFromProps(props, state) {
    if (state.src !== props.src) {
      return {
        imageError: false,
        imageLoaded: false,
        src: props.src,
      };
    }
    return null;
  }

  componentWillUnmount() {
    this.image = null;
  }

  getStyles = () => {
    const {
      animationDuration,
      aspectRatio,
      color,
      disableTransition,
      imageStyle,
      style,
    } = this.props;

    const imageTransition = !disableTransition && {
      opacity: this.state.imageLoaded ? 1 : 0,
      filterBrightness: this.state.imageLoaded ? 100 : 0,
      filterSaturate: this.state.imageLoaded ? 100 : 20,
      transition: `
        filterBrightness ${animationDuration *
          0.75}ms cubic-bezier(0.4, 0.0, 0.2, 1),
        filterSaturate ${animationDuration}ms cubic-bezier(0.4, 0.0, 0.2, 1),
        opacity ${animationDuration / 2}ms cubic-bezier(0.4, 0.0, 0.2, 1)
      `,
    };

    const styles = {
      root: {
        backgroundColor: color,
        paddingTop: `calc(${aspectRatio} * 100%)`,
        ...style,
      },
      image: {
        ...imageTransition,
        ...imageStyle,
      },
    };

    return styles;
  };

  handleLoadImage = () => {
    this.setState(state => ({ ...state, imageLoaded: true }));
    if (this.props.onLoad) {
      this.props.onLoad();
    }
  };

  handleImageError = () => {
    if (this.props.src) {
      this.setState(state => ({ ...state, imageError: true }));
    }
    if (this.props.onError) {
      this.props.onError();
    }
  };

  recursiveCloneChildren = (children: React.ReactNode | React.ReactNodeArray) =>
    React.Children.map(children, child => {
      if (!React.isValidElement(child)) return child;
      // const props = { ...{ className: '' }, ...child.props };
      let childProps = {} as ChildProps;
      if (React.isValidElement(child)) {
        if (child.type === 'picture') {
          childProps = {
            className: s.image,
          };
        }
        if (child.type === 'img') {
          childProps = {
            alt: this.props.alt,
            decoding: 'async',
            onError: this.handleImageError,
            onLoad: this.handleLoadImage,
            ref: this.image,
            src: this.props.src,
            style: this.getStyles().image,
            toggle: this.toggle,
          };
        }
      }
      childProps.children = this.recursiveCloneChildren(child.props.children);
      return React.cloneElement(child, childProps);
    });

  render() {
    const styles = this.getStyles();

    // const { imageError, imageLoaded } = this.state;
    const {
      alt,
      aspectRatio,
      children,
      className,
      color,
      disableError,
      disableSpinner,
      disableTransition,
      errorIcon,
      imageStyle,
      style,
      onClick,
      ...image
    } = this.props;
    const { loading } = this.props;

    // if (!loading) {
    //   loading = <CircularProgress id={`imageLoader--${this.loaderId}`} />;
    // }

    delete image.animationDuration;

    if (children) {
      return (
        <div
          className={cn(s.root, className)}
          onClick={onClick}
          role="presentation"
          style={styles.root}
        >
          {this.recursiveCloneChildren(children)}
          <div className={s.container}>
            <div className={s.spinner}>
              {!disableSpinner &&
                !this.state.imageLoaded &&
                !this.state.imageError &&
                loading}
              {!disableError && this.state.imageError && errorIcon}
            </div>
          </div>
        </div>
      );
    }

    return (
      <div
        className={s.root}
        onClick={onClick}
        role="presentation"
        style={styles.root}
      >
        {image.src && (
          <img
            {...image}
            alt={alt}
            decoding="async"
            className={s.image}
            ref={this.image}
            style={styles.image}
            onLoad={this.handleLoadImage}
            onError={this.handleImageError}
          />
        )}
        <div className={s.container}>
          <div className={s.spinner}>
            {!disableSpinner &&
              !this.state.imageLoaded &&
              !this.state.imageError &&
              loading}
            {!disableError && this.state.imageError && errorIcon}
          </div>
        </div>
      </div>
    );
  }
}

// Image.defaultProps = {
//   animationDuration: 3000,
//   aspectRatio: 1,
//   children: null,
//   color: '#fff',
//   crossOrigin: 'anonymous',
//   disableError: false,
//   disableSpinner: false,
//   disableTransition: false,
//   // errorIcon: <FontIcon className={s.errorIcon}>broken_image</FontIcon>,
//   errorIcon: null,
//   imageStyle: {},
//   loading: null,
//   onClick: null,
//   style: {},
// };

export default Image;
