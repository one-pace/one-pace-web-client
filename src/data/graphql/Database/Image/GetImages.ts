import { PrismaClient } from '@prisma/client';

import { AST, buildSelect } from '../../utils';

export const schema = [
  `
  type Image {
    alt:             String
    arc:             Arc
    episode:         Episode
    id:              ID!
    src:             String!
    type:            String!
    width:           String!
  }
`,
];

export const queries = [
  `
  # Retrieves all images from the database
  databaseGetAllImages: [Image]

  # Retrieves one image from the database
  databaseGetImage(
    id: ID

    src: String
  ): Image
`,
];

export const resolvers = {
  RootQuery: {
    async databaseGetAllImages(
      _parent: any,
      _args: any,
      { prisma }: { prisma: PrismaClient },
      ast: AST,
    ) {
      const select = buildSelect(ast);

      const getAllImages = await prisma.image.findMany({ select });

      return getAllImages;
    },
    async databaseGetImage(
      _parent: any,
      args: { id?: string; src?: string },
      { prisma }: { prisma: PrismaClient },
      ast: AST,
    ) {
      const select = buildSelect(ast);

      const getImage = await prisma.image.findOne({
        select,
        where: { ...args },
      });

      return getImage;
    },
  },
};
