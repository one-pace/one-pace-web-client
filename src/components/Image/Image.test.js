/* eslint-env jest */
/* eslint-disable jsx-a11y/alt-text */

import React from 'react';
import renderer from 'react-test-renderer';
// import { configure, mount } from 'enzyme';
// import Adapter from 'enzyme-adapter-react-16';

import Image from './Image';
import logo from '../Header/Lodus-logo-final.svg';
import nate1x from '../RapportSection/profile-nate-hobi-lg.png';
import nate1xW from '../RapportSection/profile-nate-hobi-lg.webp';
import nate2x from '../RapportSection/profile-nate-hobi-lg@2x.png';
import nate2xW from '../RapportSection/profile-nate-hobi-lg@2x.webp';
import nate3x from '../RapportSection/profile-nate-hobi-lg@3x.png';
import nate3xW from '../RapportSection/profile-nate-hobi-lg@3x.webp';
import nate4x from '../RapportSection/profile-nate-hobi-lg@4x.png';
import nate4xW from '../RapportSection/profile-nate-hobi-lg@4x.webp';

// configure({ adapter: new Adapter() });

describe('Image', () => {
  test('renders correctly by itself', () => {
    const image = renderer
      .create(<Image alt="Logo" src={logo} height="70" width="180" />)
      .toJSON();

    expect(image).toMatchSnapshot();
  });

  test('renders correctly with children', () => {
    const image = renderer
      .create(
        <Image alt="Nate Hobi" color="transparent" src={nate1x}>
          <picture>
            <source
              srcSet={`${nate1xW} 1x, ${nate2xW} 2x, ${nate3xW} 3x, ${nate4xW} 4x`}
              type="image/webp"
            />
            <img
              srcSet={`${nate2x} 2x, ${nate3x} 3x, ${nate4x} 4x`}
              type="image/png"
            />
          </picture>
        </Image>,
      )
      .toJSON();

    expect(image).toMatchSnapshot();
  });
});
