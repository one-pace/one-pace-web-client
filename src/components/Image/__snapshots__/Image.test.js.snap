// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image renders correctly by itself 1`] = `
<div
  className="root"
  onClick={null}
  role="presentation"
  style={
    Object {
      "backgroundColor": "#fff",
      "paddingTop": "calc(1 * 100%)",
    }
  }
>
  <img
    alt="Logo"
    className="image"
    crossOrigin="anonymous"
    decoding="async"
    height="70"
    loading={null}
    onError={[Function]}
    onLoad={[Function]}
    src="Lodus-logo-final.svg"
    style={
      Object {
        "filterBrightness": 0,
        "filterSaturate": 20,
        "opacity": 0,
        "transition": "
        filterBrightness 2250ms cubic-bezier(0.4, 0.0, 0.2, 1),
        filterSaturate 3000ms cubic-bezier(0.4, 0.0, 0.2, 1),
        opacity 1500ms cubic-bezier(0.4, 0.0, 0.2, 1)
      ",
      }
    }
    width="180"
  />
  <div
    className="container"
  >
    <div
      className="spinner"
    >
      <svg
        aria-valuemax={100}
        aria-valuemin={0}
        className="md-progress md-progress--circular md-block-centered md-progress--circular-indeterminate"
        height={24}
        id="imageLoader--Lodus-logo-final.svg"
        role="progressbar"
        style={Object {}}
        viewBox="0 0 66 66"
        width={24}
      >
        <circle
          className="md-circular-progress-path md-circular-progress-path--animated"
          cx="33"
          cy="33"
          r="30"
          strokeLinecap="round"
          strokeWidth="6"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Image renders correctly with children 1`] = `
<div
  className="root"
  onClick={null}
  role="presentation"
  style={
    Object {
      "backgroundColor": "transparent",
      "paddingTop": "calc(1 * 100%)",
    }
  }
>
  <picture
    className="image"
  >
    <source
      srcSet="profile-nate-hobi-lg.webp 1x, profile-nate-hobi-lg@2x.webp 2x, profile-nate-hobi-lg@3x.webp 3x, profile-nate-hobi-lg@4x.webp 4x"
      type="image/webp"
    />
    <img
      alt="Nate Hobi"
      decoding="async"
      onError={[Function]}
      onLoad={[Function]}
      src="profile-nate-hobi-lg.png"
      srcSet="profile-nate-hobi-lg@2x.png 2x, profile-nate-hobi-lg@3x.png 3x, profile-nate-hobi-lg@4x.png 4x"
      style={
        Object {
          "filterBrightness": 0,
          "filterSaturate": 20,
          "opacity": 0,
          "transition": "
        filterBrightness 2250ms cubic-bezier(0.4, 0.0, 0.2, 1),
        filterSaturate 3000ms cubic-bezier(0.4, 0.0, 0.2, 1),
        opacity 1500ms cubic-bezier(0.4, 0.0, 0.2, 1)
      ",
        }
      }
      type="image/png"
    />
  </picture>
  <div
    className="container"
  >
    <div
      className="spinner"
    >
      <svg
        aria-valuemax={100}
        aria-valuemin={0}
        className="md-progress md-progress--circular md-block-centered md-progress--circular-indeterminate"
        height={24}
        id="imageLoader--profile-nate-hobi-lg.png"
        role="progressbar"
        style={Object {}}
        viewBox="0 0 66 66"
        width={24}
      >
        <circle
          className="md-circular-progress-path md-circular-progress-path--animated"
          cx="33"
          cy="33"
          r="30"
          strokeLinecap="round"
          strokeWidth="6"
        />
      </svg>
    </div>
  </div>
</div>
`;
